### PR TITLE
Fix Parquet predicate pushdown for smallint, tinyint

### PR DESCRIPTION
**Pull request by nezihyigitbasi**
Backports https://github.com/prestosql/presto/pull/131
Original author is @findepi ~~couldn't retain authorship while applying the patch, but added the original author to all commit messages.~~

### CODE REVIEW

**Comment by nezihyigitbasi** [comment]
@arhimondr @wenleix do you have some cycles to take a look at this one?

**Comment by arhimondr** [comment]
@nezihyigitbasi I can get to this only next week.

File: presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java [hunk 1]
```diff
@@ -78,42 +78,63 @@ public boolean matches(long numberOfRows, Map<ColumnDescriptor, Statistics<?>> s
...
-                domain = Domain.all(type);
             }
             else {
-                domain = getDomain(type, numberOfRows, columnStatistics, id, column.toString(), failOnCorruptedParquetStatistics);
+                Domain domain = getDomain(type, numberOfRows, columnStatistics, id, column.toString(), failOnCorruptedParquetStatistics);
+                if (domain.isNone()) {
+                    return false;
+                }
+                Domain effectivePredicateDomain = effectivePredicateDomains.get(column);
+                if (effectivePredicateDomain != null) {
```
Comment (highker) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java]:
```java
if (effectivePredicateDomain != null && effectivePredicateDomain.intersect(domain).isNone()) {

File: presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java [hunk 2]
```diff
@@ -78,42 +78,63 @@ public boolean matches(long numberOfRows, Map<ColumnDescriptor, Statistics<?>> s
...
         for (RichColumnDescriptor column : columns) {
             DictionaryDescriptor dictionaryDescriptor = dictionaries.get(column);
             Domain domain = getDomain(getPrestoType(effectivePredicate, column), dictionaryDescriptor);
             if (domain != null) {
-                domains.put(column, domain);
+                if (domain.isNone()) {
+                    return false;
+                }
+                Domain effectivePredicateDomain = effectivePredicateDomains.get(column);
+                if (effectivePredicateDomain != null) {
```
Comment (highker) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java]: Same as abovr

**Comment by arhimondr** [comment]
> Use known Presto Type instead of reconstructing

This commit looks different in both sources. Is this a merge resolution artifact? Or was it changed deliberately?

File: presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java (hunk 1 above)
Comment (arhimondr) [reply presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java]: I would suggest to keep the original commit as is, and than if needed address the nits in a separate commit

File: presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java [hunk 3]
```diff
@@ -160,8 +161,10 @@ public static Domain getDomain(Type type, long rowCount, Statistics<?> statistic
             if (hasFalseValues) {
                 return Domain.create(ValueSet.of(type, false), hasNullValue);
             }
+            // All nulls case is handled earlier
+            throw new VerifyException("Impossible boolean statistics");
         }
-        else if ((type.equals(BIGINT) || type.equals(TINYINT) || type.equals(SMALLINT) || type.equals(INTEGER)) && (statistics instanceof LongStatistics || statistics instanceof IntStatistics)) {
+        if ((type.equals(BIGINT) || type.equals(TINYINT) || type.equals(SMALLINT) || type.equals(INTEGER)) && (statistics instanceof LongStatistics || statistics instanceof IntStatistics)) {
```
Comment (arhimondr) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java]: > Remove redundant `else` 

nit: original commit had a space here

**Comment by nezihyigitbasi** [comment]
> This commit looks different in both sources. Is this a merge resolution artifact? Or was it changed deliberately? If this is deliberate change, could you please extract it to a separate commit?

Removed those additional (minor/refactoring) changes.

**Pull request merged**